### PR TITLE
Info providers and devices packages

### PR DESCRIPTION
**Pull request by DmytroLinkin**
With a growing codebase grows `resource/` package, which mostly used for keeping the code which doesn't fit into other packages. To cope with that create `infoprovider/` and `devices/` packages to:

- make plugin more structured;
- allow re-usage of the code by multiple plugins;
- cover code with a new unit-tests.

`infoprovider/` package contains info providers, which are moved from `resources/` and, to have consistency, extracted from `netdevice/` package.

`devices/` package contains `rdma` and `vdpa` devices moved from `netdevice/` package. As an accompanying changes implemented `RdmaProvider` interface to allow mocking of `rdmamap` package for unit tests.
`pciDevice` is also moved into `devices/`, but its responsibility is splitted into:

- `apiDevice` - implements `APIDevice` interface to expose device specific information to Kubernetes API;
- `hostDevice` - implements `HostDevice` interface to reflect common for all devices functionality; is a base class used across the plugin;
- `pciDevice` - implements `PciDevice` interface to reflect common for all PCI devices functionality.

Common functionality extracted from `PciDevice` is a preparation for future non-PCI based devices, like Auxiliary devices.

### CODE REVIEW

File: pkg/accelerator/accelDeviceProvider.go [hunk 1]
```diff
@@ -48,8 +48,8 @@ func (ap *accelDeviceProvider) GetDiscoveredDevices() []*ghw.PCIDevice {
 	return ap.deviceList
 }
 
-func (ap *accelDeviceProvider) GetDevices(rc *types.ResourceConfig) []types.PciDevice {
-	newPciDevices := make([]types.PciDevice, 0)
+func (ap *accelDeviceProvider) GetDevices(rc *types.ResourceConfig) []types.HostDevice {
+	newPciDevices := make([]types.HostDevice, 0)
```
Comment (moshe010) [inline pkg/accelerator/accelDeviceProvider.go:52]: newHostDevices

File: pkg/devices/host.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+					{ID: int64(nodeNum)},
+				},
+			}
+			apiDevice.GetAPIDevice().Topology = topology
+		}
+	}
+
+	return &hostDevice{
+		APIDevice:     apiDevice,
+		basePciDevice: dev,
```
Comment (moshe010) [inline pkg/devices/host.go:66]: I am trying to understand why you need the basePCIDevice in the hostDevice.
if I understand the refactor the pci is save in the GenPCIDevice right?
Are you keeping it to get the vendor and product id?

File: pkg/devices/host.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+func (hd *hostDevice) GetVendor() string {
+	return hd.basePciDevice.Vendor.ID
+}
+
+func (hd *hostDevice) GetDeviceCode() string {
+	return hd.basePciDevice.Product.ID
+}
+
+func (hd *hostDevice) GetDeviceID() string {
+	return hd.GetAPIDevice().ID
```
Comment (moshe010) [inline pkg/devices/host.go:80]: What is the ID, PCI?

File: pkg/factory/factory_test.go [hunk 4]
```diff
@@ -186,11 +188,13 @@ var _ = Describe("Factory", func() {
...
 						On("GetDriver").Return(drivers[i]).
 						On("GetPciAddr").Return(pciAddr[i]).
+						On("GetDeviceID").Return(pciAddr[i]).
 						On("GetPFName").Return(pfNames[i]).
 						On("GetPfPciAddr").Return(rootDevices[i]).
 						On("GetAPIDevice").Return(&pluginapi.Device{}).
 						On("GetLinkType").Return(linkTypes[i]).
-						On("GetDDPProfiles").Return(ddpProfiles[i])
+						On("GetDDPProfiles").Return(ddpProfiles[i]).
+						On("GetVFID").Return(-1)
```
Comment (moshe010) [inline pkg/factory/factory_test.go:197]: why this added?

File: pkg/infoprovider/rdmaInfoProvider.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		return nil
+	}
+	return rip.rdmaSpec.GetRdmaDeviceSpec()
+}
+
+func (rip *rdmaInfoProvider) GetEnvVal() string {
+	return ""
+}
+
+func (rip *rdmaInfoProvider) GetMounts() []*pluginapi.Mount {
```
Comment (moshe010) [inline pkg/infoprovider/rdmaInfoProvider.go:57]: Is there real use in the GetMounts? also other return 
"
mounts := make([]*pluginapi.Mount, 0)
	return mounts
"

File: pkg/infoprovider/rdmaInfoProvider.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	return &rdmaInfoProvider{
+		rdmaSpec: rdmaSpec,
+	}
+}
+
+// *****************************************************************
+/* DeviceInfoProvider Interface */
+
+func (rip *rdmaInfoProvider) GetDeviceSpecs() []*pluginapi.DeviceSpec {
+	if !rip.rdmaSpec.IsRdma() {
```
Comment (moshe010) [inline pkg/infoprovider/rdmaInfoProvider.go:46]: should we move this to is isHealthy as done in vdpa?

File: pkg/infoprovider/vhostNetInfoProvider.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+	return deviceSpec
+}
+
+func (rip *vhostNetInfoProvider) GetEnvVal() string {
+	return ""
+}
+
+func (rip *vhostNetInfoProvider) GetMounts() []*pluginapi.Mount {
+	return nil
```
Comment (moshe010) [inline pkg/infoprovider/vhostNetInfoProvider.go:100]: same comment as for the RDMA provider

File: pkg/netdevice/pciNetDevice_test.go [hunk 8]
```diff
@@ -247,13 +165,12 @@ var _ = Describe("PciNetDevice", func() {
 					"sys/bus/pci/devices/0000:00:00.1",
 					"sys/kernel/iommu_groups/0",
 					"sys/bus/pci/drivers/vfio-pci",
-					"dev/vhost-net",
+					// "dev/vhost-net", FIXME: /dev folder is not mocked
```
Comment (moshe010) [inline pkg/netdevice/pciNetDevice_test.go:168]: why this change?

File: pkg/accelerator/accelDevice_test.go [hunk 9]
```diff
@@ -43,7 +43,6 @@ var _ = Describe("Accelerator", func() {
 					Files: map[string][]byte{"sys/bus/pci/devices/0000:00:00.1/numa_node": []byte("0")},
 				}
 				defer fs.Use()()
-				utils.SetDefaultMockNetlinkProvider()
```
[the remark this answers is not in the document]
Comment (DmytroLinkin) [reply pkg/accelerator/accelDevice_test.go:46]: Netlink is not invoked for Accelerator anywhere

File: pkg/accelerator/accelDeviceProvider.go (hunk 1 above)
Comment (DmytroLinkin) [reply pkg/accelerator/accelDeviceProvider.go:52]: Ok

File: pkg/devices/devices.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+import (
+	"github.com/golang/glog"
+	"github.com/jaypipes/ghw"
+
+	"github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/utils"
+)
+
+// GenNetDevice is a generic network device embedded into top level devices
+type GenNetDevice struct {
```
[the remark this answers is not in the document]
Comment (DmytroLinkin) [reply pkg/devices/devices.go:25]: Ok

File: pkg/devices/host.go (hunk 2 above)
Comment (DmytroLinkin) [reply pkg/devices/host.go:66]: Yes, it is for vendor and product ID.

File: pkg/devices/host.go (hunk 3 above)
Comment (DmytroLinkin) [reply pkg/devices/host.go:80]: For `Accelerator` and `NetPciDevice` it is PCI address, for `AuxNetDevice` it is auxiliary name, eg `mlx5_core.sf.3`

File: pkg/factory/factory_test.go (hunk 4 above)
Comment (DmytroLinkin) [reply pkg/factory/factory_test.go:197]: Test panics without it.

File: pkg/infoprovider/rdmaInfoProvider.go (hunk 5 above)
Comment (DmytroLinkin) [reply pkg/infoprovider/rdmaInfoProvider.go:57]: It is used at `(ad *apiDevice) GetMounts()`. Can't say anything about return value. I've just moved the code.

File: pkg/infoprovider/rdmaInfoProvider.go (hunk 6 above)
Comment (DmytroLinkin) [reply pkg/infoprovider/rdmaInfoProvider.go:46]: Not sure. `IsRdma()` is a part of RDMA Spec and it return true if RDMA devices were discovered, while `isHealthy` is a part of info provider and verifies if provided vdpa type is supported. Mb in both cases not correct name was chosen for the function.

File: pkg/infoprovider/vhostNetInfoProvider.go (hunk 7 above)
Comment (DmytroLinkin) [reply pkg/infoprovider/vhostNetInfoProvider.go:100]: same answer

File: pkg/netdevice/pciNetDevice_test.go (hunk 8 above)
Comment (DmytroLinkin) [reply pkg/netdevice/pciNetDevice_test.go:168]: To emphasize that this path is useless, since it is not mocked.

File: pkg/infoprovider/vdpaInfoProvider.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+		return nil
+	}
+	devSpecs := make([]*pluginapi.DeviceSpec, 0)
+
+	// DeviceSpecs only required for vhost vdpa type as the
+	if vip.vdpaType == types.VdpaVhostType {
+		devSpecs = append(devSpecs, &pluginapi.DeviceSpec{
+			HostPath:      vip.dev.GetPath(),
+			ContainerPath: vip.dev.GetPath(),
+			Permissions:   "mrw",
```
Comment (e0ne) [inline pkg/infoprovider/vdpaInfoProvider.go:58]: Since we do refactoring, maybe it's better to move this values to the constant?

File: pkg/infoprovider/vdpaInfoProvider.go [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+		})
+	}
+	return devSpecs
+}
+
+// GetEnvVal returns the environment variable value
+func (vip *vdpaInfoProvider) GetEnvVal() string {
```
Comment (e0ne) [inline pkg/infoprovider/vdpaInfoProvider.go:65]: Are `GetEnvVal()` used anywhere? They always return only empty string

File: pkg/infoprovider/vhostNetInfoProvider.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+// NewVhostNetInfoProvider returns a new Vhost Information Provider
+func NewVhostNetInfoProvider() types.DeviceInfoProvider {
+	return &vhostNetInfoProvider{}
+}
+
+// VhostNetDeviceExist returns true if /dev/vhost-net exists
+func VhostNetDeviceExist() bool {
+	_, err := os.Stat("/dev/vhost-net")
```
Comment (e0ne) [inline pkg/infoprovider/vhostNetInfoProvider.go:42]: Actually, this method could return false for different reasons. Please check if error is ErrNotExist or rename the method

File: pkg/infoprovider/vhostNetInfoProvider.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+		HostPath:      "/dev/vhost-net",
+		ContainerPath: "/dev/vhost-net",
+		Permissions:   "mrw",
+	})
+
+	return deviceSpec
+}
+
+// TunDeviceExist returns true if /dev/net/tun exists
+func tunDeviceExist() bool {
```
Comment (e0ne) [inline pkg/infoprovider/vhostNetInfoProvider.go:59]: The same comment as for VhostNetDeviceExist

File: pkg/devices/api.go [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+func (ad *apiDevice) GetEnvVal() string {
+	// Currently Device Plugin does not support returning multiple Env Vars
+	// so we use the value provided by the first InfoProvider.
+	return ad.infoProviders[0].GetEnvVal()
+}
+
+func (ad *apiDevice) GetMounts() []*pluginapi.Mount {
+	mnt := make([]*pluginapi.Mount, 0)
+	for _, infoProvider := range ad.infoProviders {
+		mnt = append(mnt, infoProvider.GetMounts()...)
```
Comment (e0ne) [inline pkg/devices/api.go:60]: Do we need to check if `GetMounts` returns nil?

**Review by e0ne — CHANGES_REQUESTED** [review]

File: pkg/infoprovider/rdmaInfoProvider.go [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+func NewRdmaInfoProvider(rdmaSpec types.RdmaSpec) types.DeviceInfoProvider {
+	return &rdmaInfoProvider{
+		rdmaSpec: rdmaSpec,
+	}
+}
+
+// *****************************************************************
+/* DeviceInfoProvider Interface */
+
+func (rip *rdmaInfoProvider) GetDeviceSpecs() []*pluginapi.DeviceSpec {
```
[the remark this answers is not in the document]
Comment (DmytroLinkin) [reply pkg/infoprovider/rdmaInfoProvider.go:45]: I could change it to `ip` or `r`

File: pkg/infoprovider/vdpaInfoProvider.go (hunk 11 above)
Comment (DmytroLinkin) [reply pkg/infoprovider/vdpaInfoProvider.go:58]: Should it be constant? Asking, because don't know.

File: pkg/infoprovider/vdpaInfoProvider.go (hunk 12 above)
Comment (DmytroLinkin) [reply pkg/infoprovider/vdpaInfoProvider.go:65]: `GetEnvVal()` called at `(ad *apiDevice) GetMounts()`, before this change it was call in `PciDevice`. `GetMounts()` in turn called at `(rp *ResourcePoolImpl) GetMounts(deviceIDs []string) []*pluginapi.Mount` which is a part of KAPI plugin must implement.

File: pkg/infoprovider/vhostNetInfoProvider.go (hunk 13 above)
Comment (DmytroLinkin) [reply pkg/infoprovider/vhostNetInfoProvider.go:42]: Frankly, this functions aren't really usable, except that they just hide call to `Stat`. Mb it's better to call it directly?

File: pkg/infoprovider/vhostNetInfoProvider.go (hunk 14 above)
Comment (DmytroLinkin) [reply pkg/infoprovider/vhostNetInfoProvider.go:59]: Same answer here

File: pkg/devices/api.go (hunk 15 above)
Comment (DmytroLinkin) [reply pkg/devices/api.go:60]: For what? `append` works correctly with nil slice.

**Review by e0ne — APPROVED** [review]
LGTM

File: pkg/devices/host.go (hunk 2 above)
Comment (moshe010) [reply pkg/devices/host.go:66]: so why NewHostDevice need the full ghw.PCIDevice? You can pass just the vendor and Product ID

File: pkg/netdevice/pciNetDevice_test.go (hunk 8 above)
Comment (moshe010) [reply pkg/netdevice/pciNetDevice_test.go:168]: so the path was useless before the refactoring ?

File: pkg/netdevice/pciNetDevice.go [hunk 17]
```diff
@@ -55,81 +53,48 @@ func NewPciNetDevice(dev *ghw.PCIDevice, rFactory types.ResourceFactory, rc *typ
...
-	netDevs, _ := utils.GetNetNames(pciAddr)
-	if len(netDevs) == 0 {
-		ifName = ""
-	} else {
-		ifName = netDevs[0]
-	}
-	pfName, err := utils.GetPfName(pciAddr)
+	pciDev, err := devices.NewGenPciDevice(dev)
 	if err != nil {
-		glog.Warningf("unable to get PF name %q", err.Error())
```
Comment (moshe010) [inline pkg/netdevice/pciNetDevice.go:90]: why you removed the warning?

File: pkg/netdevice/pciNetDevice_test.go (hunk 8 above)
Comment (DmytroLinkin) [reply pkg/netdevice/pciNetDevice_test.go:168]: Yes. For the testing purposes it make sense to implement it in the future

**Review by moshe010 — APPROVED** [review]
LGTM

**Comment by DmytroLinkin** [comment]
Accidantly closed PR

**Comment by DmytroLinkin** [comment]
@adrianchiris are we good to merge this PR?

**Comment by adrianchiris** [comment]
@DmytroLinkin  could you rebase this PR. 

planning to get to it this week (fingers crossed :) )

**Review by SchSeba — APPROVED** [review]
/lgtm

I also run the sriov-operator functional tests with this image of the device plugin and all green

```
Ran 25 of 27 Specs in 2148.218 seconds
SUCCESS! -- 25 Passed | 0 Failed | 0 Pending | 2 Skipped
```

File: pkg/devices/host.go [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	// Use the default Information Provided if not
+	if len(infoProviders) == 0 {
+		infoProviders = append(infoProviders, rFactory.GetDefaultInfoProvider(deviceID, driverName))
+	}
+
+	apiDevice := NewAPIDevice(deviceID, infoProviders)
+
+	if !rc.ExcludeTopology {
```
Comment (adrianchiris) [inline pkg/devices/host.go:53]: can this be handled in `NewAPIDevice` ?

File: pkg/netdevice/pciNetDevice_test.go [hunk 19]
```diff
@@ -71,80 +80,6 @@ var _ = Describe("PciNetDevice", func() {
 				Expect(dev.GetAPIDevice().Topology.Nodes[0].ID).To(Equal(int64(0)))
 				Expect(err).NotTo(HaveOccurred())
 			})
-			It("should not populate topology due to negative numa_node", func() {
```
Comment (adrianchiris) [inline pkg/netdevice/pciNetDevice_test.go:74]: note to self, moved to host_test.go

File: pkg/netdevice/pciNetDevice_test.go [hunk 20]
```diff
@@ -267,35 +208,14 @@ var _ = Describe("PciNetDevice", func() {
...
-				Expect(dev.GetDeviceSpecs()).To(HaveLen(4)) // /dev/vfio/vfio0 and default /dev/vfio/vfio + vhost-net + tun
+				Expect(dev.GetDeviceSpecs()).To(HaveLen(4)) // /dev/vfio/0 and default /dev/vfio/vfio + vhost-net + tun
 				Expect(dev.GetRdmaSpec().IsRdma()).To(BeFalse())
 				Expect(dev.GetRdmaSpec().GetRdmaDeviceSpec()).To(HaveLen(0))
 				Expect(dev.GetLinkType()).To(Equal(""))
 				Expect(dev.GetAPIDevice().Topology.Nodes[0].ID).To(Equal(int64(0)))
 				Expect(err).NotTo(HaveOccurred())
 			})
 		})
-		Context("cannot get device's driver", func() {
```
Comment (adrianchiris) [inline pkg/netdevice/pciNetDevice_test.go:278]: note to self, moved to `host_test.go`

File: pkg/devices/host.go (hunk 18 above)
Comment (adrianchiris) [reply pkg/devices/host.go:53]: that is just get nodeNum here and let APIDevice handle the creation of Topology WDYT?

File: pkg/accelerator/accelDevice_test.go [hunk 21]
```diff
@@ -26,6 +27,16 @@ import (
 	. "github.com/onsi/gomega"
 )
 
+var pciAddr = "0000:00:00.1"
```
Comment (adrianchiris) [inline pkg/accelerator/accelDevice_test.go:30]: not a fan of globals in this case.

can you move pciAddr to be within Describe statement and define newPciDevice inline within Describe as well to be used by the tests.

Comment (adrianchiris) [reply pkg/accelerator/accelDevice_test.go:30]: e.g

```
Describe("some thing", func() {
    pciAddr := "00000:00:00.1"
    newPciDeviceFn := func() *ghw.PciDevice { return &ghw.PciDevice{....} }

}
```

File: pkg/types/types.go [hunk 22]
```diff
@@ -184,6 +195,24 @@ type DeviceProvider interface {
 	ValidConfig(*ResourceConfig) bool
 }
 
+// Device interfaces relationship:
```
Comment (adrianchiris) [inline pkg/types/types.go:198]: please remove this ascii drawing from source code as its clear from the interfaces below who embeds who.
also if we keep it we need to maintain it.

File: pkg/types/types.go [hunk 23]
```diff
@@ -201,23 +230,29 @@ type HostDevice interface {
...
 type PciDevice interface {
 	HostDevice
 	GetPciAddr() string
+}
+
+// NetDevice provides an interface to get generic network device information
+type NetDevice interface {
+	HostDevice
+	GetPFName() string
 	GetPfPciAddr() string
```
Comment (adrianchiris) [inline pkg/types/types.go:244]: can you move `GetPfPciAddr` and `GetVFID` to `PciDevice` ?

we can say that if a PciDevice does not have a parent PCI Address or a VF ID it will return empty string or -1 respectively WDYT ?


Also with all this refactoring and addition of interface i think we would want to add docstring to all methods so its clear what each method does.

File: pkg/types/types.go [hunk 24]
```diff
@@ -201,23 +230,29 @@ type HostDevice interface {
...
 // extends HostDevice interface
 type PciDevice interface {
 	HostDevice
 	GetPciAddr() string
+}
+
+// NetDevice provides an interface to get generic network device information
+type NetDevice interface {
+	HostDevice
+	GetPFName() string
```
Comment (adrianchiris) [inline pkg/types/types.go:243]: can you rename to `GetPFNetName` ? it seems more suitable.

**Comment by adrianchiris** [comment]
Took a while to get to this one, appreciate the patience @DmytroLinkin ! 
added some comments. overall i think it looks good.

File: pkg/resources/deviceSelectors.go [hunk 25]
```diff
@@ -185,7 +188,7 @@ func getItem(hay []string, needle string) string {
 	return ""
 }
 
-func isSelected(dev types.PciDevice, selector string) bool {
+func isSelected(id int, selector string) bool {
```
[the remark this answers is not in the document]
Comment (DmytroLinkin) [reply pkg/resources/deviceSelectors.go:191]: It will in the Aux.Devs support PR

File: pkg/types/types.go (hunk 23 above)
Comment (DmytroLinkin) [reply pkg/types/types.go:244]: `AuxNetDevice` from AuxDevs PR use it, but `AccelDevice` don't. Keep it at `NetDevice` or move to `HostDevice`?

File: pkg/types/types.go (hunk 24 above)
Comment (DmytroLinkin) [reply pkg/types/types.go:243]: Should it be `GetPFNetName` or `GetPfNetName` as `GetPfPciAddr`? Or make `PF` everywhere?

File: pkg/resources/deviceSelectors.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (adrianchiris) [reply pkg/resources/deviceSelectors.go:191]: so, `deviceIndex` ? or `devIdx` ?

File: pkg/types/types.go (hunk 24 above)
Comment (adrianchiris) [reply pkg/types/types.go:243]: +1 with being consistent. i dont have a special preference.

File: pkg/types/types.go (hunk 23 above)
Comment (adrianchiris) [reply pkg/types/types.go:244]: OK, looking at aux dev PR i see that GetVFID is renamed to a more generic GetFuncID.
And GetPfPciAddr() is actually only used in the context of a Network device selectors.

lets keep as is.


> Also with all this refactoring and addition of interface i think we would want to add docstring to all methods so its clear what each method does.

lets just add docstrings then

File: pkg/devices/api.go [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+	"github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/types"
+)
+
+type apiDevice struct {
+	device        *pluginapi.Device
+	infoProviders []types.DeviceInfoProvider
+}
+
+// NewAPIDevice returns an interface of type APIDevice
+func NewAPIDevice(id string, infoProviders []types.DeviceInfoProvider) types.APIDevice {
```
[the remark this answers is not in the document]
Comment (DmytroLinkin) [reply pkg/devices/api.go:32]: What this give us?

[the remark this answers is not in the document]
Comment (adrianchiris) [reply pkg/devices/api.go:32]: it just common practice, to return (pointer to) struct and not interface. 

_"accept interfaces return structs"_

1. generally these new methods are regarded as constructors, IMO a constructor should construct a concrete type
2. Extending the implementation with public methods that are not part of the interface would be problematic to test by tests that reside in their own "_test" package, as well as using them in general.  (also generally a struct's bound methods may implement multiple interfaces as well).

**Review by SchSeba — APPROVED** [review]

**Review by adrianchiris — APPROVED** [review]
LGTM ! Thanks for working on this one @DmytroLinkin
